File: app/src/main/java/me/aluceps/sandbox/view/main/MainPresenter.java
package me.aluceps.sandbox.view.main;

import android.support.annotation.NonNull;

import javax.inject.Inject;

import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import me.aluceps.sandbox.model.ConnpassEvent;
import me.aluceps.sandbox.repository.ConnpassRepository;

public class MainPresenter implements MainContract.Presenter<MainContract.View> {

    private MainContract.View view;

    private ConnpassRepository repository;

    private CompositeDisposable disposable;

    @Inject
    public MainPresenter(@NonNull ConnpassRepository repository,
                         @NonNull CompositeDisposable disposable) {
        this.repository = repository;
        this.disposable = disposable;
    }

    @Override
    public void setView(MainContract.View view) {
        this.view = view;
    }

    @Override
    public void load(final boolean isRefresh) {
        if (!view.checkConnectionState()) {
            view.hideProgressBar(isRefresh);
            return;
        }
        view.showProgressBar(isRefresh);
        repository.events()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<ConnpassEvent>() {
                    @Override
                    public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {
                        disposable.add(d);
                        view.clear();
                    }

                    @Override
                    public void onSuccess(@io.reactivex.annotations.NonNull ConnpassEvent connpassEvent) {
                        view.hideProgressBar(isRefresh);
                        view.setEvents(connpassEvent.getEvents());
                    }

                    @Override
                    public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                    }
                });
    }

    @Override
    public void destroy() {
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }
}

File: app/src/main/java/me/aluceps/sandbox/view/main/MainPresenter.kt
package me.aluceps.sandbox.view.main

import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import me.aluceps.sandbox.model.ConnpassEvent
import me.aluceps.sandbox.model.RequestParams
import me.aluceps.sandbox.repository.ConnpassRepository
import javax.inject.Inject

class MainPresenter @Inject
constructor(private val repository: ConnpassRepository,
            private val disposable: CompositeDisposable) : MainContract.Presenter<MainContract.View> {

    private lateinit var view: MainContract.View

    private val requestParams: RequestParams = RequestParams()

    private var loading: Boolean = false

    init {
        requestParams.limit = RequestParams.LIMIT
        requestParams.offset = 0
        requestParams.total = 1
    }

    override fun setView(view: MainContract.View) {
        this.view = view
    }

    override fun load() {
        view.showProgressBar()
        if (!loading && requestParams.total > requestParams.offset) {
            repository.events(requestParams)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(object : SingleObserver<ConnpassEvent> {
                        override fun onSubscribe(@io.reactivex.annotations.NonNull d: Disposable) {
                            disposable.add(d)
                            loading = true
                        }

                        override fun onSuccess(@io.reactivex.annotations.NonNull connpassEvent: ConnpassEvent) {
                            view.hideProgressBar()
                            view.clear()
                            connpassEvent.events?.let {
                                view.setData(it)
                            }
                            requestParams.offset = connpassEvent.resultsStart + connpassEvent.resultsReturned
                            requestParams.total = connpassEvent.resultsAvailable
                            loading = false
                        }

                        override fun onError(@io.reactivex.annotations.NonNull e: Throwable) {
                            view.hideProgressBar()
                            loading = false
                        }
                    })
        }
    }

    override fun refresh() {
        view.clear()
        requestParams.offset = 0
        requestParams.total = 1
        load()
    }

    override fun isLoading(): Boolean = loading

    override fun destroy() {
        if (disposable != null && !disposable.isDisposed) {
            disposable.dispose()
        }
    }
}
